### PR TITLE
[cpp_on_shore_task_2] vector learning

### PR DESCRIPTION
**Pull request by SheldonLoveCoding**
### What?

- brief: 阅读STL vector的源码，学习vector的使用和测试
- author: liuxd_sheldon@163.com

### Directory Specification

```
├─cpp_on_shore
    │  ├─chris
    │  │  ├─doc
    │  │  │  └─task_2
    │  │  └─src
    │  │      └─task_2
    │  ├─roger_lv
    │  │  ├─doc
    │  │  │  └─task_2
    │  │  └─src
    │  │      └─task_2
    │  └─sheldon
    │      ├─doc
    │      │  └─task_2
    │      └─src
    │          └─task_2
```

### What changes are included in this PR?

按照要求创建task_2学习目录，分别创建doc和src文件夹，保存task_2各自笔记和代码

### Are these changes tested?

编写的代码经过测试

### CODE REVIEW

File: stl/cpp_on_shore/chris/src/task_2/my_vector.h [hunk 1]
```diff
@@ -0,0 +1,148 @@
+#include <cstring>
+#include <iostream>
+
+template <typename T>
+class My_Vector {
```
Comment (Light-City) [inline stl/cpp_on_shore/chris/src/task_2/my_vector.h:5]: follow google style

File: stl/cpp_on_shore/roger_lv/src/task_2/my_test.h [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+  for (int i = 0; i < 100000; ++i) {
+    vc.emplace_back(Person(10, "roger"));
+  }
+  end = std::chrono::steady_clock::now();
+  std::cout << "Test the emplace time: "
+            << std::chrono::duration_cast<std::chrono::nanoseconds>(end - start).count()
+            << " ns" << std::endl;
+}
+
+void TestAboutAt(std::vector<Person>& vc) {
```
Comment (Light-City) [inline stl/cpp_on_shore/roger_lv/src/task_2/my_test.h:39]: may be google test frame is a good idea

File: stl/cpp_on_shore/roger_lv/src/task_2/my_test.h [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+ public:
+  MyVector() {
+    data = new T[10];
+    size = 0;
+    capacity = 10;
+  }
+  MyVector(int n) {
+    data = new T[n];
+    for (int i = 0; i < n; ++i) {
+      data[i] = T();
```
Comment (Light-City) [inline stl/cpp_on_shore/roger_lv/src/task_2/my_test.h:75]: this line is unsafe, consider: T() = delete

File: stl/cpp_on_shore/roger_lv/src/task_2/my_test.h [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+  ~MyVector() { delete[] data; }
+  void push_back(const T& x) {
+    if (size == capacity) {
+      capacity *= 2;
+      T* temp = new T[capacity];
+      for (int i = 0; i < size; ++i) {
+        temp[i] = data[i];
+      }
+      temp[++size] = x;
+      data = temp;
```
Comment (Light-City) [inline stl/cpp_on_shore/roger_lv/src/task_2/my_test.h:90]: please free old memory

File: stl/cpp_on_shore/roger_lv/src/task_2/my_test.h [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+  void pop_back() {
+    if (size == 0) {
+      std::cout << "The vector is empty!" << std::endl;
+      return;
+    }
+    --size;
+  }
+  T& operator[](int index) {
+    if (index < 0 || index >= size) {
+      std::cout << "The index is out of range!" << std::endl;
```
Comment (Light-City) [inline stl/cpp_on_shore/roger_lv/src/task_2/my_test.h:104]: throw exception

File: stl/cpp_on_shore/roger_lv/src/task_2/my_test.h [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+      return data[0];
+    }
+    return data[index];
+  }
+  T& Size() { return size; }
+  T& Capacity() { return capacity; }
```
Comment (Light-City) [inline stl/cpp_on_shore/roger_lv/src/task_2/my_test.h:110]: why return T&?

File: stl/cpp_on_shore/roger_lv/src/task_2/my_test.h (hunk 2 above)
Comment (Rogerlv51) [reply stl/cpp_on_shore/roger_lv/src/task_2/my_test.h]: I'll learn it later, but now my os has some problems sry.

File: stl/cpp_on_shore/roger_lv/src/task_2/my_test.h (hunk 4 above)
Comment (Rogerlv51) [reply stl/cpp_on_shore/roger_lv/src/task_2/my_test.h:89]: already done

**Comment by Light-City** [comment]
lgtm

**Pull request merged**